controller/controller_server/csi_controller_server.py

```diff
@@ -534,6 +539,83 @@ def GetCapacity(self, request, context):
+
+            try:
+                array_type, volume_id = utils.get_volume_id_info(request.volume_id)
+            except ObjectIdError as ex:
+                logger.warning("volume id is invalid. error : {}".format(ex))
+                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                return csi_pb2.ControllerExpandVolumeResponse()
+
+            with get_agent(user, password, array_addresses, array_type).get_mediator() as array_mediator:
+                logger.debug(array_mediator)
+
+                size = request.capacity_range.required_bytes
+
+                if size > array_mediator.maximal_volume_size_in_bytes:
+                    raise controller_errors.SizeOutOfRange(size, array_mediator.maximal_volume_size_in_bytes)
+
+                try:
+
+                    logger.debug("expanding volume {0}".format(volume_id))
+                    array_mediator.expand_volume(
+                        volume_id=volume_id,
+                        required_bytes=request.capacity_range.required_bytes)
+                    res = utils.generate_csi_expand_volume_response(request)
+                    logger.info("finished expanding volume")
+                    return res
+
+                except controller_errors.PermissionDeniedError as ex:
+                    context.set_code(grpc.StatusCode.PERMISSION_DENIED)
+                    context.set_details(ex)
+                    return csi_pb2.ControllerExpandVolumeResponse()
+
+        except controller_errors.ObjectNotFoundError as ex:
+            logger.info("Volume not found: {0}".format(ex))
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+            context.set_details(ex)
+            return csi_pb2.ControllerExpandVolumeResponse()
+
+        except controller_errors.ObjectIsStillInUseError as ex:
+            logger.info("Could not expand volume while in use: {0}".format(ex))
+            context.set_code(grpc.StatusCode.FAILED_PRECONDITION)
+            context.set_details(ex.message)
+            return csi_pb2.ControllerExpandVolumeResponse()
+
+        except (ValidationException, controller_errors.IllegalObjectID) as ex:
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            return csi_pb2.ControllerExpandVolumeResponse()
+
+        except controller_errors.NotEnoughSpaceInPool as ex:
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.RESOURCE_EXHAUSTED)
+            return csi_pb2.ControllerExpandVolumeResponse()
+
+        except controller_errors.SizeOutOfRange as ex:
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.OUT_OF_RANGE)
+            return csi_pb2.ControllerExpandVolumeResponse()
+
+        except (controller_errors.ObjectIsStillInUseError, Exception) as ex:
```

Comment:
done

Reply:
why `FAILED_PRECONDITION`? I thought `ObjectIsStillInUseError` is for still being used by a flashcopy mapping, not a worker node (since we support online expansion).

Reply:
yesyes, my bad